### PR TITLE
Package Manager uses TabControl to be more Accessible

### LINKED ISSUE #9078: Accessibility issues in Screen Reader experiences related to "browse","install","update" Tabs
Various VS Feedback accessibility issues relating to the Tabs in the Package Manager UI:

- [x] After pressing on "Browse" tab focus is not navigating inside the tab  (**1049188**)

- [x] Narrator is reading "1 of 1" for "browse","install","update" tabs (**1049208**)

- [x] Narrator is announcing incorrect state for the non selected tab item. (**994450**)

### LINKED ISSUE #9192: [Test Failure] There is an error “The calling thread cannot access this object because a different thread owns it” after installing/updating/uninstalling package in the PM UI
## Details about Problem
NuGet version: dev-donnie-msft-a11yPmuiTabs\5.6.0.10272
VS Version: Master\29818.127
OS: Windows-10-Enterprise-RS5
## Detailed repro steps so we can see the same problem
1. Launch VS and create a project [e.g. Console App (.NET Core)].
2. Open the PM UI of the project or the solution.
3. Install/Update/Uninstall a package in the PM UI.
4. Observe the Error List window.
## Expected
There is no error in the Error List window.
## Actual
There is an error “The calling thread cannot access this object because a different thread owns it” after installing/updating/uninstalling package in the PM UI as below screenshot.
image

## Note
1.	The error will disappear after building the project/solution.
2.	The operation of installing/updating/uninstalling package is successful even the error appears.

### PR DESCRIPTION
**Pull request by donnie-msft**
## Bug

Fixes: https://github.com/NuGet/Home/issues/9078
Fixes: https://github.com/NuGet/Home/issues/9192 (bug during this branch only)
Regression: No  

## Fix

**Purpose of this work**: There were 3 accessibility issues with the Package Manager Window (PMUI). The approach here was refactoring of XAML controls to get some of the native accessibility support offered there, rather than trying to force accessibility with our nested controls.

Tabs should look the same (eg, here's the new tabs on Solution PMUI):
image

**Accessibility bugs**:
1. After pressing on "Browse" tab focus is not navigating inside the tab

     - **Resolution**: TabControl works by allowing arrow keys to navigate tabItems. The <Tab> key will now navigate into the controls within the page.

2. Narrator is reading "1 of 1" for "browse","install","update" tabs
     - **Resolution**: Narrator now recognizes a proper ItemsControl list.

3. Narrator is announcing incorrect state for the non selected tab item. 
     - **Resolution**: TabControl doesn't separate the Focus from pressing <Tab> from the Selected Item. Therefore, this problem has gone away. 

**Code Change Summary**:

- Tabs are now a Tab Control
-  Removed FilterLabel custom control.
-  Fewer controls are now being created (all FilterLabels contained a Warning Icon and Count text block, but were always hidden on some tabs). 
- Consolidate Tab is now dynamically created in code behind. 
     - Why?.... Showing/Hiding a TabItem proved challenging because it showed a tiny tab with no content, and Narrator could still recognize that tab and counted it in the Total tab Count. 
     - Creating it only for Solution Package Manager saves it from being created in the Project Package Manager window unnecessarily.
          - Code could be improved from a maintenance standpoint. Potentially utilizing more XAML ControlTemplates (or similar) would reduce code redundancy. Follow-up issue: https://github.com/NuGet/Home/issues/9183

**Known Issue** with WPF causing Narrator to read Tab Headers twice sometimes. I will be following-up with an email thread about this to ensure we are encountering a known bug.

## Testing/Validation

Tests Added: No  
Reason for not adding tests:  Logically, this should only be a refactoring. Existing tests should still make sense.

**Validation**:  

- Launching PMUI in Project and Solution mode, made sure Warning Icon and Package Count showed up where appropriate.

- Listened to Narrator read the Tab headers. Tested the <Tab> key to ensure it moved into the correct area.

- Requesting support from our manual test Team to look as well. 

- Accessibility Insights passed tests... Example:
image

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Controls/TabItemButton.cs [hunk 1]
```diff
@@ -22,4 +22,4 @@ public void Select()
             OnClick();
         }
     }
-}
\ No newline at end of file
+}
```
Comment (rrelyea) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Controls/TabItemButton.cs]: there is no significant change in this file. 
options:
a) keep the change in...end up with cr-lf at end
b) undo the change via "git checkout dev -- <PATHTOFILE>", i believe -- so the file is not changed

**Comment by rrelyea** [comment]
believe you have a diff in detailcontrol.xaml - likely a merge conflict with my change that went in over the weekend. Please make sure to take my change (used labelby)

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml [hunk 2]
```diff
@@ -155,4 +155,4 @@
       </Border>
     </DockPanel>
   </ScrollViewer>
-</UserControl>
\ No newline at end of file
+</UserControl>
```
Comment (rrelyea) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml]: Same feedback as earlier. no significant change on this file.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs [hunk 3]
```diff
@@ -143,7 +142,7 @@ internal PackageSourceMoniker SelectedSource
 
             SetTitle();
 
-            _topPanel.IsSolution = Model.IsSolution;
+            _topPanel.CreateAndAddConsolidateTab(Model.IsSolution);
```
Comment (rrelyea) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: this method probably doesn't need to be called if IsSolution is false. (make sure topPanel has right default for IsSolution in case it doesn't get called.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs [hunk 4]
```diff
@@ -763,7 +762,7 @@ internal async Task SearchPackagesAndRefreshUpdateCountAsync(string searchText,
                 {
                     // it means selected tab is update itself, so just wait for searchAsyncTask to complete
                     // without making another call to loader to get all packages.
-                    _topPanel._labelUpgradeAvailable.Count = 0;
+                    _topPanel.ShowCountOnUpdatesTab(count: 0);
```
Comment (rrelyea) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: Perhaps "UpdateCountOnUpdatesTab(int count)" would be better naming?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs [hunk 5]
```diff
@@ -787,8 +786,8 @@ private void RefreshInstalledAndUpdatesTabs()
             {
                 await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
 
-                _topPanel._labelInstalled.ShowWarning = false;
-                _topPanel._labelUpgradeAvailable.Count = 0;
+                _topPanel.ShowWarningOnInstalledTab(installedDeprecatedPackagesCount: 0);
```
Comment (rrelyea) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: Perhaps "UpdateDeprecatedCountOnInstalledTab(int count)" would be better naming?   -- or something like that...Show sounds like it will always be shown.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs [hunk 6]
```diff
@@ -857,13 +850,13 @@ private void RefreshConsolidatablePackagesCount()
                 NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async () =>
                 {
                     await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-                    _topPanel._labelConsolidate.Count = 0;
+                    _topPanel.ShowCountOnConsolidateTab(count: 0);
```
Comment (rrelyea) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: similar naming feedback

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml [hunk 7]
```diff
@@ -25,47 +27,136 @@
...
-        Margin="35,0,0,0"
-        Text="{x:Static nuget:Resources.Action_Consolidate}"
-        ControlSelected="FilterLabel_ControlSelected"
-        Visibility="Collapsed" />
+      <TabControl x:Name="tabsPackageManagement" IsSynchronizedWithCurrentItem="False" SelectionChanged="TabsPackageManagement_SelectionChanged">
+        <TabControl.Resources>
+          <Style TargetType="TabItem">
+            <Setter Property="Background" Value="{DynamicResource {x:Static nuget:Brushes.HeaderBackground}}" />
+            <Setter Property="FontSize" Value="{Binding FontSize, RelativeSource={RelativeSource AncestorType=UserControl},Converter={StaticResource Font122PercentSizeConverter}}"/>
+            <Setter Property="Foreground" Value="{DynamicResource {x:Static nuget:Brushes.UIText}}" />
```
Comment (rrelyea) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml:39]: FontSize and Foreground setters may not be doing anything?? (don't see any uses inside the controltemplate below via TemplateBinding) -- could be wrong...should test with and without

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml [hunk 8]
```diff
@@ -25,47 +27,136 @@
...
+                      <RowDefinition Height="auto" />
+                    </Grid.RowDefinitions>
+                    <Border x:Name="mainBorder" Background="{TemplateBinding Background}" BorderThickness="0" Margin="0">
+                    </Border>
+                    <ContentPresenter  Grid.Row="0" x:Name="contentPresenter" ContentSource="Header" Focusable="False" HorizontalAlignment="{Binding HorizontalContentAlignment, RelativeSource={RelativeSource AncestorType={x:Type ItemsControl}}}" Margin="{TemplateBinding Padding}" RecognizesAccessKey="True" SnapsToDevicePixels="{TemplateBinding SnapsToDevicePixels}" VerticalAlignment="{Binding VerticalContentAlignment, RelativeSource={RelativeSource AncestorType={x:Type ItemsControl}}}"/>
+                    <!-- the line under the text to indicate that this label is selected -->
+                    <Rectangle
+                      x:Name="_underline"
+                      Grid.Row="1"
+                      Margin="0"
```
Comment (rrelyea) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml]: Margin default value is 0, so this is extraneous and (very) slightly wasteful.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml (hunk 7 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml:39]: since the foreground triggers below likely work, i am likely wrong about ForegRound. maybe about fontsize too. could be useful since they are inheritable properties.

**Review by rrelyea — CHANGES_REQUESTED** [review]
Not done with review...but have to run to meeting.
I left off at tabBrowse element inside of PackageManagerTopPanel.xaml

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs [hunk 9]
```diff
@@ -13,14 +18,145 @@ namespace NuGet.PackageManagement.UI
...
+        }
+
+        public void CreateAndAddConsolidateTab(bool isSolution)
+        {
+            IsSolution = isSolution;
+            if (IsSolution)
+            {
+                TabItem tabConsolidate = new TabItem();
+                tabConsolidate.Name = nameof(tabConsolidate);
+                tabConsolidate.Tag = "Consolidate"; //From ItemFilter Enum.
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]:
```
private static readonly string Consolidate = ItemFilter.Consolidate.ToString();
…
tabConsolidate.Tag = Consolidate;
```
Then remove the comment.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs [hunk 10]
```diff
@@ -13,14 +18,145 @@ namespace NuGet.PackageManagement.UI
...
 
         public PackageManagerTopPanel()
         {
             InitializeComponent();
+        }
+
+        public void CreateAndAddConsolidateTab(bool isSolution)
+        {
+            IsSolution = isSolution;
+            if (IsSolution)
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: You can eliminate a lot of indentation by short-circuiting for the opposite condition:
```C#
if (!IsSolution)
{
    return;
}

…
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs [hunk 11]
```diff
@@ -13,14 +18,145 @@ namespace NuGet.PackageManagement.UI
...
+                };
+
+                TextBlock textConsolidate = new TextBlock()
+                {
+                    Name = nameof(textConsolidate),
+                    Text = Resx.Action_Consolidate
+                };
+                sp.Children.Add(textConsolidate);
+
+                SetConsolidationAutomationProperties(tabConsolidate, 0);
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: Use named parameter to explain that mysterious `0`.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs [hunk 12]
```diff
@@ -13,14 +18,145 @@ namespace NuGet.PackageManagement.UI
...
         {
             InitializeComponent();
+        }
+
+        public void CreateAndAddConsolidateTab(bool isSolution)
+        {
+            IsSolution = isSolution;
+            if (IsSolution)
+            {
+                TabItem tabConsolidate = new TabItem();
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: Here and below you can use `var` for `new` initializations.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs [hunk 13]
```diff
@@ -13,14 +18,145 @@ namespace NuGet.PackageManagement.UI
...
+
+                SetConsolidationAutomationProperties(tabConsolidate, 0);
+
+                //The textblock that displays the count.
+                _countConsolidateContainer = new Border()
+                {
+                    Name = nameof(_countConsolidateContainer),
+                    CornerRadius = new CornerRadius(2),
+                    Margin = new Thickness(3, 0, 3, 0),
+                    Padding = new Thickness(3, 0, 3, 0),
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: Use named parameters.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs [hunk 14]
```diff
@@ -13,14 +18,145 @@ namespace NuGet.PackageManagement.UI
...
+                };
+                _countConsolidateContainer.SetResourceReference(Border.BackgroundProperty, Brushes.TabPopupBrushKey);
+                sp.Children.Add(_countConsolidateContainer);
+
+                _countConsolidate = new TextBlock()
+                {
+                    Name = nameof(_countConsolidate),
+                    HorizontalAlignment = HorizontalAlignment.Right,
+                    VerticalAlignment = VerticalAlignment.Top,
+                   // FontSize = new Binding( ="{Binding FontSize, RelativeSource={RelativeSource AncestorType=UserControl}}"
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: Remove commented-out code.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs [hunk 15]
```diff
@@ -13,14 +18,145 @@ namespace NuGet.PackageManagement.UI
...
+        }
+
+        private void SetConsolidationAutomationProperties(TabItem tabConsolidate, int count)
+        {
+            string automationString = null;
+            if (count > 0)
+            {
+                automationString = string.Format(
+                    CultureInfo.CurrentCulture,
+                    " {0}{1}",
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: Why is there a leading space?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs [hunk 16]
```diff
@@ -13,14 +18,145 @@ namespace NuGet.PackageManagement.UI
...
+        public void ShowWarningOnInstalledTab(int installedDeprecatedPackagesCount)
+        {
+            bool hasInstalledDeprecatedPackages = installedDeprecatedPackagesCount > 0;
+            if (hasInstalledDeprecatedPackages)
+            {
+                _warningIcon.Visibility = Visibility.Visible;
+                _warningIcon.ToolTip = string.Format(
+                        CultureInfo.CurrentCulture,
+                        NuGet.PackageManagement.UI.Resources.Label_Installed_DeprecatedWarning,
+                        installedDeprecatedPackagesCount);
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: Indent only 4 spaces from previous line.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs [hunk 17]
```diff
@@ -772,7 +771,7 @@ internal async Task SearchPackagesAndRefreshUpdateCountAsync(string searchText,
                         IncludePrerelease = IncludePrerelease
                     };
 
-                    _topPanel._labelUpgradeAvailable.Count = Model.CachedUpdates.Packages.Count;
+                    _topPanel.ShowCountOnUpdatesTab(count: Model.CachedUpdates.Packages.Count);
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: The named parameter isn't necessary here.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs [hunk 18]
```diff
@@ -820,7 +813,7 @@ private void RefreshInstalledAndUpdatesTabs()
                     IncludePrerelease = IncludePrerelease
                 };
 
-                _topPanel._labelUpgradeAvailable.Count = Model.CachedUpdates.Packages.Count;
+                _topPanel.ShowCountOnUpdatesTab(count: Model.CachedUpdates.Packages.Count);
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: Named parameter isn't necessary here.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs [hunk 19]
```diff
@@ -857,13 +850,13 @@ private void RefreshConsolidatablePackagesCount()
...
                     await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-                    _topPanel._labelConsolidate.Count = 0;
+                    _topPanel.ShowCountOnConsolidateTab(count: 0);
                     var loadContext = new PackageLoadContext(ActiveSources, Model.IsSolution, Model.Context);
                     var packageFeed = await CreatePackageFeedAsync(loadContext, ItemFilter.Consolidate, _uiLogger);
                     var loader = new PackageItemLoader(
                         loadContext, packageFeed, includePrerelease: IncludePrerelease);
 
-                    _topPanel._labelConsolidate.Count = await loader.GetTotalCountAsync(100, CancellationToken.None);
+                    _topPanel.ShowCountOnConsolidateTab(count: await loader.GetTotalCountAsync(100, CancellationToken.None));
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: The first named parameter is not necessary, but a named parameter is necessary for the magical `100` argument.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs [hunk 20]
```diff
@@ -13,14 +18,145 @@ namespace NuGet.PackageManagement.UI
...
+                return tabsPackageManagement.SelectedItem as TabItem;
+            }
+            set
+            {
+                tabsPackageManagement.SelectedItem = value;
+            }
+        }
+        public TabItem _tabConsolidate { get; private set; }
+        public Border _countConsolidateContainer { get; private set; }
+        public TextBlock _countConsolidate { get; private set; }
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: The naming convention used here is non-standard.  These members should be named `SelectedTabItem`, `TabConsolidate`, `CountConsolidateContainer`, and `CountConsolidate`, respectively.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs [hunk 21]
```diff
@@ -127,42 +241,50 @@ public void FilterLabel_ControlSelected(object sender, EventArgs e)
...
+
+            if (previousTabItem != null)
+            {
+                ItemFilter previousFilter = GetItemFilter(previousTabItem);
+
+                if (FilterChanged != null)
+                {
+                    FilterChanged(this, new FilterChangedEventArgs(previousFilter));
+                }
+            }
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs:277]: No point in getting the item filter if there's no listener.  How about:
```C#
if (previousTabItem != null && FilterChanged != null)
{ 
    ItemFilter previousFilter = GetItemFilter(previousTabItem); 
    FilterChanged(this, new FilterChangedEventArgs(previousFilter)); 
}
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs (hunk 20 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: Here's [the XAML on Dev branch](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/FilterLabel.xaml#L54) where the same controls are prefixed with _. In a discussion, it was mentioned that WPF turns these into private fields, therefore fall under the convention of an _ prefix.

Or are you suggesting I rename the existing XAML controls as well?

Comment (dtivel) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: But they are no longer private fields with your change.  You are introducing public properties.  Rename the controls, if necessary.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml [hunk 22]
```diff
@@ -25,47 +27,136 @@
...
+          <TabItem.Header>
+            <StackPanel Orientation="Horizontal">
+              <TextBlock x:Name="textInstalled" Text="{x:Static nuget:Resources.Label_Installed}"></TextBlock>
+              <imaging:CrispImage
+                Name="_warningIcon"
+                Margin="3,0"
+                Width="13"
+                Height="13"
+                Visibility="Collapsed"
+                HorizontalAlignment="Center"
```
Comment (rrelyea) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml]: does horizontalalignment do anything here?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml [hunk 23]
```diff
@@ -25,47 +27,136 @@
...
+        <TabItem Name="tabInstalled" Tag="Installed">
+          <AutomationProperties.Name>
+            <MultiBinding StringFormat=" {0}{1}">
+              <Binding ElementName="textInstalled" Path="Text" />
+              <Binding ElementName="_warningIcon" Path="ToolTip" />
+            </MultiBinding>
+          </AutomationProperties.Name>
+          <TabItem.Header>
+            <StackPanel Orientation="Horizontal">
+              <TextBlock x:Name="textInstalled" Text="{x:Static nuget:Resources.Label_Installed}"></TextBlock>
```
Comment (rrelyea) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml]: Use:
`<TextBlock /> instead of <TextBock    ></TextBlock>`
Also on TextBlock with name of "textUpdates"

**Review by rrelyea — CHANGES_REQUESTED** [review]

**Comment by donnie-msft** [comment]
> believe you have a diff in detailcontrol.xaml - likely a merge conflict with my change that went in over the weekend. Please make sure to take my change (used labelby)

@rrelyea Must have gotten resolved somewhere. Ping me if you still see this
diff I'm seeing:
> src/NuGet.Clients/NuGet.PackageManagement.UI/NuGet.PackageManagement.UI.csproj
> src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/FilterLabel.xaml
> src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/FilterLabel.xaml.cs
> src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs
> src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml
> src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs (hunk 4 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: Went with SetCountOnUpdatesTab... Let me know if that doesn't work.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs (hunk 5 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: SetDeprecationWarningOnInstalledTab
OK?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml (hunk 7 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml:39]: Before I added Foreground, the color was always white, I think.

FontSize is needed here to utilize the Converter for scaling up the size.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml (hunk 8 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml]: More perf wins!

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs (hunk 12 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: Done w/ 1 tradeoff...

Couldn't use `nameof `with textConsolidate's inline constructor variable initialization, so had to split it out, like `textConsolidate.Name = ...`.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs (hunk 5 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: i prefer my suggession...feel free to ask another teammate which they prefer. the fact that count is passed in, and that when it is 0 means you are not always "setting a deprecation warning".
your proposal is an improvement over current code.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml (hunk 7 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml:39]: i was suggesting it may be worthwhile to confirm they are needed.

**Comment by rrelyea** [comment]
> @rrelyea Must have gotten resolved somewhere. 
Yes. I think it is better now.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs (hunk 15 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: In the equivalent code in XAML, it seems that the XAML parser can't understand "{0}", which is why we have always had a leading space there.
eg: ([from dev](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/FilterLabel.xaml#L30))
`<MultiBinding StringFormat=" {0}{1}">`

xaml.cs parses this, but I kept the space for consistency. I can test and see if it breaks. Didn't see any value in having a different format string in code-behind (which has to be maintained as a special case).

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs (hunk 5 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: We are thinking to leave it as I had it with this mindset:

A method named `EnableWarningIcon(bool isEnabled)` would be equivalent. I fully expect that passing a `false` here in fact disables it. My logic was... I can accept the Count argument and make the decision inside the method (essentially, `bool isWarningShown`) without the caller having to decide in advance.

Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: Random example I found trying to find similar naming and conventions...

Windows Forms has [Graphics.FillRectangle ](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.graphics.fillrectangle?view=netframework-4.8#System_Drawing_Graphics_FillRectangle_System_Drawing_Brush_System_Drawing_Rectangle_) which takes a brush. That brush could be [Brushes.Transparent](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.brushes.transparent?view=netframework-4.8). Technically this still fills the rectangle, but not visibly.... not a perfect comparison.

Fill is a nice word. More descriptive than Show. Just not sure what else to use here. Any way to integrate the word `Visibility`? ...

Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs]: Compromise: `UpdateDeprecationStatusOnInstalledTab`

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml (hunk 22 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml]: I observed that it can slightly change how the image is drawn, depending on the container's width/height. However, I used a magnifier on our icon, and saw no change with Horizontal/Vertical alignment, so I will remove them.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml (hunk 7 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml:39]: They are. I rechecked, and the font size/color looks incorrect:

image

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs (hunk 15 above)
Comment (rrelyea) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: To escape a markup extension in XAML use :
`   someProp="{}{0}{1}"`

**Review by rrelyea — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs (hunk 15 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: Changed to Rob's suggestion.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs (hunk 9 above)
Comment (donnie-msft) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs]: I just reverted this as I'm no longer using strings.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs [hunk 24]
```diff
@@ -13,14 +18,138 @@ namespace NuGet.PackageManagement.UI
...
         public PackageManagerTopPanel()
         {
             InitializeComponent();
+        }
+
+        public void CreateAndAddConsolidateTab()
+        {
+            var tabConsolidate = new TabItem();
+            tabConsolidate.Name = nameof(tabConsolidate);
+            tabConsolidate.Tag = ItemFilter.Consolidate;
```
Comment (dtivel) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerTopPanel.xaml.cs:45]: Nice!

**Review by dtivel — APPROVED** [review]

**Comment by donnie-msft** [comment]
Threading bug was introduced during one of my commits, but is already fixed. Notes here: https://github.com/NuGet/Home/issues/9192
